File: src/com/l2kt/gameserver/network/serverpackets/Die.java
package com.l2kt.gameserver.network.serverpackets;

import com.l2kt.gameserver.data.manager.CastleManager;
import com.l2kt.gameserver.model.actor.Attackable;
import com.l2kt.gameserver.model.actor.Creature;
import com.l2kt.gameserver.model.actor.instance.Player;
import com.l2kt.gameserver.model.entity.Siege;
import com.l2kt.gameserver.model.pledge.Clan;

public class Die extends L2GameServerPacket
{
	private final Creature _activeChar;
	private final int _charObjId;
	private final boolean _fake;
	
	private boolean _sweepable;
	private boolean _allowFixedRes;
	private Clan _clan;
	
	public Die(Creature cha)
	{
		_activeChar = cha;
		_charObjId = cha.getObjectId();
		_fake = !cha.isDead();
		
		if (cha instanceof Player)
		{
			Player player = (Player) cha;
			_allowFixedRes = player.getAccessLevel().allowFixedRes();
			_clan = player.getClan();
			
		}
		else if (cha instanceof Attackable)
			_sweepable = ((Attackable) cha).isSpoiled();
	}
	
	@Override
	protected final void writeImpl()
	{
		if (_fake)
			return;
		
		writeC(0x06);
		writeD(_charObjId);
		writeD(0x01); // to nearest village
		
		if (_clan != null)
		{
			Siege.SiegeSide side = null;
			
			final Siege siege = CastleManager.getInstance().getActiveSiege(_activeChar);
			if (siege != null)
				side = siege.getSide(_clan);
			
			writeD((_clan.hasHideout()) ? 0x01 : 0x00); // to clanhall
			writeD((_clan.hasCastle() || side == Siege.SiegeSide.OWNER || side == Siege.SiegeSide.DEFENDER) ? 0x01 : 0x00); // to castle
			writeD((side == Siege.SiegeSide.ATTACKER && _clan.getFlag() != null) ? 0x01 : 0x00); // to siege HQ
		}
		else
		{
			writeD(0x00); // to clanhall
			writeD(0x00); // to castle
			writeD(0x00); // to siege HQ
		}
		
		writeD((_sweepable) ? 0x01 : 0x00); // sweepable (blue glow)
		writeD((_allowFixedRes) ? 0x01 : 0x00); // FIXED
	}
}

File: src/com/l2kt/gameserver/network/serverpackets/Die.kt
package com.l2kt.gameserver.network.serverpackets

import com.l2kt.gameserver.data.manager.CastleManager
import com.l2kt.gameserver.model.actor.Attackable
import com.l2kt.gameserver.model.actor.Creature
import com.l2kt.gameserver.model.actor.instance.Player
import com.l2kt.gameserver.model.entity.Siege
import com.l2kt.gameserver.model.pledge.Clan

class Die(private val _activeChar: Creature) : L2GameServerPacket() {
    private val _charObjId: Int = _activeChar.objectId
    private val _fake: Boolean = !_activeChar.isDead

    private var _sweepable: Boolean = false
    private var _allowFixedRes: Boolean = false
    private var _clan: Clan? = null

    init {

        if (_activeChar is Player) {
            _allowFixedRes = _activeChar.accessLevel.allowFixedRes
            _clan = _activeChar.clan

        } else if (_activeChar is Attackable)
            _sweepable = _activeChar.isSpoiled
    }

    override fun writeImpl() {
        if (_fake)
            return

        writeC(0x06)
        writeD(_charObjId)
        writeD(0x01) // to nearest village

        if (_clan != null) {
            var side: Siege.SiegeSide? = null

            val siege = CastleManager.getInstance().getActiveSiege(_activeChar)
            if (siege != null)
                side = siege.getSide(_clan)

            writeD(if (_clan!!.hasHideout()) 0x01 else 0x00) // to clanhall
            writeD(if (_clan!!.hasCastle() || side == Siege.SiegeSide.OWNER || side == Siege.SiegeSide.DEFENDER) 0x01 else 0x00) // to castle
            writeD(if (side == Siege.SiegeSide.ATTACKER && _clan!!.flag != null) 0x01 else 0x00) // to siege HQ
        } else {
            writeD(0x00) // to clanhall
            writeD(0x00) // to castle
            writeD(0x00) // to siege HQ
        }

        writeD(if (_sweepable) 0x01 else 0x00) // sweepable (blue glow)
        writeD(if (_allowFixedRes) 0x01 else 0x00) // FIXED
    }
}